File: Semestre_2/Vision_Computadora/Entregas/Matlab/T2_2_Histograma_2D.m
close all;
clc;

[file1,path1] = uigetfile({'*.png';'*.jpg';'*.jpeg';'*.bmp';'*.gif'},'File Selector');

ruta1=strcat(path1,file1);

A=imread(ruta1);
A=im2double(A);

% Canales RGB (convertidos a 0-255)
A_R = A(:,:,1) * 255;
A_G = A(:,:,2) * 255;
A_B = A(:,:,3) * 255;

num_puntos = 600;
idx = randperm(numel(A_R), min(num_puntos, numel(A_R))); 
R = A_R(idx);
G = A_G(idx);
B = A_B(idx);

figure('Color', 'white', 'Position', [100, 100, 1200, 400]);
subplot(141); imshow(A);title('Imagen original');

% Histograma: Canal Rojo vs Canal Verde
subplot(142);scatter(R, G, 10, 'k', 'filled');
xlabel('Canal Rojo (0-255)', 'Color', [1 0 0], 'FontWeight', 'bold');
ylabel('Canal Verde (0-255)', 'Color', [0 0.5 0], 'FontWeight', 'bold');
title('Canales R y G', 'FontSize', 12);
grid on;
xlim([0 255]); ylim([0 255]);
ax = gca;
ax.XColor = [1 0 0];
ax.YColor = [0 0.5 0];
ax.YDir = 'reverse';

% Histograma: Canal Verde vs Canal Azul 
subplot(143); scatter(G, B, 10, 'k', 'filled');
xlabel('Canal Verde (0-255)', 'Color', [0 0.5 0], 'FontWeight', 'bold');
ylabel('Canal Azul (0-255)', 'Color', [0 0 1], 'FontWeight', 'bold');
title('Canales G y B', 'FontSize', 12);
grid on;
xlim([0 255]); ylim([0 255]);
ax = gca;
ax.XColor = [0 0.5 0];
ax.YColor = [0 0 1];
ax.YDir = 'reverse';

% Histograma: Canal Rojo vs Canal Azul
subplot(144); scatter(R, B, 10, 'k', 'filled');
xlabel('Canal Rojo (0-255)', 'Color', [1 0 0], 'FontWeight', 'bold');
ylabel('Canal Azul (0-255)', 'Color', [0 0 1], 'FontWeight', 'bold');
title('Canales R y B', 'FontSize', 12);
grid on;
xlim([0 255]); ylim([0 255]);
ax = gca;
ax.XColor = [1 0 0];
ax.YColor = [0 0 1];
ax.YDir = 'reverse';

File: Semestre_2/Vision_Computadora/Entregas/Python/T2_2_Histograma_2D.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from tkinter import Tk, filedialog

Tk().withdraw()
file_path = filedialog.askopenfilename(
    initialdir="C:\\Users\\death\\Documents\\Maestria\\MCIA\\Semestre_2\\Vision_Computadora\\img",
    filetypes=[("Images", "*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)

A = imread(file_path)

if A.shape[-1] == 4:
    A = A[..., :3]

A_f = A.astype(float) / 255.0

A_R = A_f[..., 0] * 255
A_G = A_f[..., 1] * 255
A_B = A_f[..., 2] * 255

num_puntos = 600
idx = np.random.permutation(A_R.size)[:num_puntos]

R = A_R.flatten()[idx]
G = A_G.flatten()[idx]
B = A_B.flatten()[idx]

plt.figure(figsize=(12,4))
plt.subplot(1,4,1); plt.imshow(A/255); plt.title('Imagen original'); plt.axis('off')

plt.subplot(1,4,2); plt.scatter(R, G, s=10, c='k')
plt.xlabel('Canal Rojo (0-255)', color='red')
plt.ylabel('Canal Verde (0-255)', color='green')
plt.title('Canales R y G')
plt.grid(True)
plt.xlim(0,255); plt.ylim(0,255)
ax = plt.gca()
ax.spines['bottom'].set_color('red')
ax.spines['left'].set_color('green')
ax.xaxis.label.set_weight('bold')
ax.yaxis.label.set_weight('bold')
ax.invert_yaxis()

plt.subplot(1,4,3); plt.scatter(G, B, s=10, c='k')
plt.xlabel('Canal Verde (0-255)', color='green')
plt.ylabel('Canal Azul (0-255)', color='blue')
plt.title('Canales G y B')
plt.grid(True)
plt.xlim(0,255); plt.ylim(0,255)
ax = plt.gca()
ax.spines['bottom'].set_color('green')
ax.spines['left'].set_color('blue')
ax.xaxis.label.set_weight('bold')
ax.yaxis.label.set_weight('bold')
ax.invert_yaxis()

plt.subplot(1,4,4); plt.scatter(R, B, s=10, c='k')
plt.xlabel('Canal Rojo (0-255)', color='red')
plt.ylabel('Canal Azul (0-255)', color='blue')
plt.title('Canales R y B')
plt.grid(True)
plt.xlim(0,255); plt.ylim(0,255)
ax = plt.gca()
ax.spines['bottom'].set_color('red')
ax.spines['left'].set_color('blue')
ax.xaxis.label.set_weight('bold')
ax.yaxis.label.set_weight('bold')
ax.invert_yaxis()

plt.show()
